File: redblacktree/redblacktree.go
package redblacktree

const (
	RED   = 0
	BLACK = 1
)

// Node struct contains actual data, color and address to left, right and
// parent node
type Node struct {
	Data   int
	Color  int
	Parent *Node
	Left   *Node
	Right  *Node
}

// Tree struct contains the pointer to root node and total number of nodes.
type RBTree struct {
	Root *Node
	Size int
}

// NewRBTree creates an empty Red Black Tree
func NewRBTree() *RBTree {
	return &RBTree{nil, 0}
}

// NewNode creates a new Node with default integer data
func NewNode(data, color int) *Node {
	return &Node{data, color, nil, nil, nil}
}

// nodeColor returns the color of node
func nodeColor(node *Node) int {
	if node == nil {
		return BLACK
	}
	return node.Color
}

// uncle returns the uncle of node
func (node *Node) uncle() *Node {
	if node == nil || node.Parent == nil || node.Parent.Parent == nil {
		return nil
	}
	return node.Parent.sibling()
}

// sibling returns the sibling of node
func (node *Node) sibling() *Node {
	if node == nil || node.Parent == nil {
		return nil
	}
	if node == node.Parent.Left {
		return node.Parent.Right
	}
	return node.Parent.Left
}

// grandparent returns the grandparent of node
func (node *Node) grandparent() *Node {
	if node == nil || node.Parent == nil {
		return nil
	}
	return node.Parent.Parent
}

File: RedBlackTree/RedBlackTree.cpp
#include <iostream>
#include "RedBlackTree.hpp"

/**
 * Delete red black tree nodes recursively
*/
void RedBlackTree::deleteTreeRecursive(Node* node) {
    if (node == nullptr)
        return;

    if (node->left != nullptr)
        deleteTreeRecursive(node->left);
    if (node->right != nullptr)
        deleteTreeRecursive(node->right);
    delete node;
}

/**
 * returns node color
 * @param node pointer to red black tree node
*/
int RedBlackTree::getNodeColor(Node* node) {
    if (node == nullptr)
        return BLACK;
    return node->color;
}

/**
 * returns uncle of node
 * @param node pointer to red black tree node
*/
Node* RedBlackTree::uncle(Node* node) {
    if (node == nullptr || node->parent == nullptr ||
        node->parent->parent == nullptr)
            return nullptr;

    return sibling(node->parent);
}

/**
 * returns sibling of given node
 * @param node pointer to red black tree node
*/
Node* RedBlackTree::sibling(Node* node) {
    if (node == nullptr || node->parent == nullptr)
        return nullptr;

    if (node == node->parent->left)
        return node->parent->right;
    return node->parent->left;
}

/**
 * returns grandparent of node
 * @param node pointer to red black tree node
*/
Node* RedBlackTree::grandparent(Node* node) {
    if (node == nullptr || node->parent == nullptr)
        return nullptr;
    return node->parent->parent;
}
